server/src/main/java/org/cloudfoundry/identity/uaa/oauth/token/UserTokenGranter.java

```diff
@@ -97,6 +104,16 @@ protected Authentication validateRequest(TokenRequest request) {
         ClientDetails receiving = clientDetailsService.loadClientByClientId(request.getRequestParameters().get(CLIENT_ID), IdentityZoneHolder.get().getId());
         super.validateGrantType(GRANT_TYPE_REFRESH_TOKEN, receiving);
 
+        //7. receiving client must have a client secret
```

Comment:
I vote to perform the check in extra method, means, check if secret "" is set to client.  I assume you agree if I provide a PR which would perform the check for you?